server/src/main/java/org/elasticsearch/index/query/WildcardQueryBuilder.java

```diff
@@ -182,14 +183,24 @@ public static WildcardQueryBuilder fromXContent(XContentParser parser) throws IO
     
     @Override
     protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
-        if ("_index".equals(fieldName)) {
-            // Special-case optimisation for canMatch phase:  
-            // We can skip querying this shard if the index name doesn't match the value of this query on the "_index" field.
-            QueryShardContext shardContext = queryRewriteContext.convertToShardContext();
-            if (shardContext != null && shardContext.indexMatches(BytesRefs.toString(value)) == false) {
+        QueryShardContext context = queryRewriteContext.convertToShardContext();
+        if (context != null) {
+            MappedFieldType fieldType = context.fieldMapper(this.fieldName);
+            if (fieldType == null) {
                 return new MatchNoneQueryBuilder();
-            }            
+            } else if (fieldType instanceof ConstantFieldType) {
+                // This logic is correct for all field types, but by only applying it to constant
+                // fields we also have the guarantee that it doesn't perform I/O, which is important
+                // since rewrites might happen on a network thread.
+                Query query = fieldType.wildcardQuery(value, null, context); // the rewrite method doesn't matter
```

Comment:
I don't think we have any fields in this case today but I wanted to be cautious because this would be a bad bug that could be undetected for a long time, and the upcoming wildcard field might have to perform I/O in order to select the ngrams that have the lowest document frequency when building an approximation in order for the approximation to not hit the maximum clause count while still being as selective as possible.